fxgcppubsub/codec/codec.go

```diff
@@ -0,0 +1,254 @@
+package codec
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/hamba/avro/v2"
+	"github.com/linkedin/goavro/v2"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/proto"
+)
+
+var _ Codec = (*DefaultCodec)(nil)
+
+// Codec is the interface for components in charge to handle raw, avro or protobuf encoding and decoding.
+type Codec interface {
+	Encode(in any) ([]byte, error)
+	Decode(enc []byte, out any) error
+}
+
+// DefaultCodec is the default Codec implementation.
+type DefaultCodec struct {
+	schemaType       pubsub.SchemaType
+	schemaEncoding   pubsub.SchemaEncoding
+	schemaDefinition string
+}
+
+// NewDefaultCodec returns a new DefaultCodec instance.
+func NewDefaultCodec(schemaType pubsub.SchemaType, schemaEncoding pubsub.SchemaEncoding, schemaDefinition string) *DefaultCodec {
+	return &DefaultCodec{
+		schemaType:       schemaType,
+		schemaEncoding:   schemaEncoding,
+		schemaDefinition: schemaDefinition,
+	}
+}
+
+// Encode encodes an input into an avro or protobuf encoded slice of bytes.
+//
+//nolint:cyclop,exhaustive
+func (c *DefaultCodec) Encode(in any) ([]byte, error) {
```

Comment:
How about splitting the codec per schema type, so if anyone wants to go with an alternative codec, they don't have to reimplement one for all schema types they don't necessarily use? Not sure if it worths the effort, just throwing in the idea

Reply:
high level publisher / subscriber completely obfuscate the fact there is auto ser/de, therefore that there is codecs involved behind.

imo overriding the codecs should be pretty rare, but if they do so, they just need to implement only the format support they need (raw, avro, proto)